### PR TITLE
HPCC-18422 Add new format meta data to indexes

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [x] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [x] Security
  - [x] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
Added code to dumpkey to examine the metadata, and checked it is appearing correctly.

### CODE REVIEW

File: roxie/ccd/ccdserver.cpp [hunk 1]
```diff
@@ -12033,6 +12034,12 @@ class CRoxieServerIndexWriteActivity : public CRoxieServerInternalSinkActivity,
             metadata->setPropBin("_record_layout", layoutMetaSize, layoutMetaBuff);
             rtlFree(layoutMetaBuff);
         }
+        if (helper.queryOutputMeta() && helper.queryOutputMeta()->queryTypeInfo())
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: This code (and the function it is in) COULD be commoned up. But that's arguably a separate issue.

File: roxie/ccd/ccdserver.cpp [hunk 2]
```diff
@@ -12132,10 +12139,7 @@ class CRoxieServerIndexWriteActivity : public CRoxieServerInternalSinkActivity,
                 }
                 reccount++;
             }
-            if(metadata)
-                builder->finish(metadata,&fileCrc);
-            else
-                builder->finish(&fileCrc);
+            builder->finish(metadata, &fileCrc);
```
Comment (richardkchapman) [inline roxie/ccd/ccdserver.cpp]: I removed the overloaded finish() method as it didn't seem to me to add any clarity or efficiency (indeed, rather the reverse).

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 3]
```diff
@@ -184,21 +142,138 @@ class ECLRTL_API CDynamicDiskReadArg : public CThorDiskReadArg
...
+    void addFilter(const char *filter)
+    {
+        filters.addFilter(filter);
+    }
+
 private:
     StringAttr fileName;
-    unsigned numOffsets = 0;
     unsigned flags = 0;
-    Owned<IRtlFieldTypeDeserializer> indeserializer;   // Owns the resulting ITypeInfo structures, so needs to be kept around
```
Comment (richardkchapman) [inline rtl/eclrtl/eclhelper_dyn.cpp]: These were unused and should have been removed in prior PR

File: ecl/hthor/hthor.cpp [hunk 4]
```diff
@@ -1209,13 +1202,22 @@ void CHThorIndexWriteActivity::execute()
...
+    // Legacy record layout info
     void * layoutMetaBuff;
     size32_t layoutMetaSize;
     if(helper.getIndexLayout(layoutMetaSize, layoutMetaBuff))
     {
         properties.setPropBin("_record_layout", layoutMetaSize, layoutMetaBuff);
         rtlFree(layoutMetaBuff);
     }
+    // New record layout info
+    if (helper.queryOutputMeta() && helper.queryOutputMeta()->queryTypeInfo())
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: This should only happen if the format is serializable.

File: ecl/hthor/hthor.cpp [hunk 5]
```diff
@@ -1280,11 +1282,11 @@ void CHThorIndexWriteActivity::buildLayoutMetadata(Owned<IPropertyTree> & metada
     metadata->setProp("_record_ECL", helper.queryRecordECL());
 
     void * layoutMetaBuff;
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: Now unused

File: ecl/hthor/hthor.cpp [hunk 6]
```diff
@@ -1280,11 +1282,11 @@ void CHThorIndexWriteActivity::buildLayoutMetadata(Owned<IPropertyTree> & metada
...
-    size32_t layoutMetaSize;
-    if(helper.getIndexLayout(layoutMetaSize, layoutMetaBuff))
+    if (helper.queryOutputMeta() && helper.queryOutputMeta()->queryTypeInfo())
```
Comment (ghalliday) [inline ecl/hthor/hthor.cpp]: only if can be serialized.

File: ecl/hthor/hthorkey.cpp [hunk 7]
```diff
@@ -95,15 +94,7 @@ IRecordLayoutTranslator * getRecordLayoutTranslator(IDefRecordMeta const * activ
     IPropertyTree const & props = df->queryAttributes();
     MemoryBuffer diskMetaBuff;
     if(!props.getPropBin("_record_layout", diskMetaBuff))
-#ifdef IGNORE_FORMAT_CRC_MISMATCH_WHEN_NO_METADATA
```
Comment (ghalliday) [inline ecl/hthor/hthorkey.cpp]: Was this an intentional change?

File: roxie/ccd/ccdserver.cpp [hunk 8]
```diff
@@ -12027,11 +12028,11 @@ class CRoxieServerIndexWriteActivity : public CRoxieServerInternalSinkActivity,
         metadata->setProp("_record_ECL", helper.queryRecordECL());
 
         void * layoutMetaBuff;
```
Comment (ghalliday) [inline roxie/ccd/ccdserver.cpp]: now unused. + following code should only be used if can serialize the type info

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 9]
```diff
@@ -106,13 +63,14 @@ class ECLRTL_API CDynamicDiskReadArg : public CThorDiskReadArg
         const char *epos = strchr(filter,'=');
         assertex(epos);
         StringBuffer fieldName(epos-filter, filter);
-        unsigned fieldNum = inrec->getFieldNum(fieldName);
+        unsigned fieldNum = inrec.getFieldNum(fieldName);
         assertex(fieldNum != (unsigned) -1);
```
Comment (ghalliday) [inline rtl/eclrtl/eclhelper_dyn.cpp]: Could do with a more helpful error message

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 10]
```diff
@@ -106,13 +63,14 @@ class ECLRTL_API CDynamicDiskReadArg : public CThorDiskReadArg
         const char *epos = strchr(filter,'=');
         assertex(epos);
```
Comment (ghalliday) [inline rtl/eclrtl/eclhelper_dyn.cpp]: needs a better error message (and also doesn't support ~)

File: rtl/eclrtl/rtldynfield.cpp [hunk 11]
```diff
@@ -30,67 +30,83 @@
 
 //---------------------------------------------------------------------------------------------------------------------
 
-const RtlTypeInfo *FieldTypeInfoStruct::createRtlTypeInfo() const
+const RtlTypeInfo *FieldTypeInfoStruct::createRtlTypeInfo(bool useBias) const
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: not sure about useBias as the name of the parameter.
- unsigned fields aren't biased
- string fields aren't translated.
- fields in the payload aren't biased.
- there are some weird rules for integers inside keyed rows - I have a feeling they don't work.

isLegacyKeyedField may be better

File: rtl/eclrtl/rtldynfield.cpp [hunk 12]
```diff
@@ -30,67 +30,83 @@
...
 {
     const RtlTypeInfo *ret = nullptr;
-    switch (fieldType & RFTMkind)
+    auto useType = fieldType;
+    if (useBias)
+    {
+#if __BYTE_ORDER == __LITTLE_ENDIAN
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: Would it be better to have a RtKeyednteger type (which was derived from RtlIntTypeInfo or RtSwaplIntTypeInfo depending on the endianness), and which implemented the bias support?

File: rtl/eclrtl/rtldynfield.cpp [hunk 13]
```diff
@@ -518,7 +543,7 @@ class CRtlFieldTypeDeserializer : public CInterfaceOf<IRtlFieldTypeDeserializer>
         if (found)
             return *found;
         info.locale = keep(info.locale);
-        const RtlTypeInfo * ret = info.createRtlTypeInfo();
+        const RtlTypeInfo * ret = info.createRtlTypeInfo(useBias);
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: Not as simple as a boolean - it depends if the field is keyed.

File: rtl/eclrtl/rtldynfield.cpp [hunk 14]
```diff
@@ -683,13 +709,41 @@ class CRtlFieldTypeDeserializer : public CInterfaceOf<IRtlFieldTypeDeserializer>
             }
         }
         info.fieldType &= ~RFTMserializerFlags;
-        return info.createRtlTypeInfo();
+        return info.createRtlTypeInfo(useBias);
+    }
+    void patchIndexFilePos()
+    {
+        if (callback && (base->fieldType & RFTMkind) == type_record)
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: base->getType() is cleaner.

File: rtl/eclrtl/rtldynfield.cpp [hunk 15]
```diff
@@ -683,13 +709,41 @@ class CRtlFieldTypeDeserializer : public CInterfaceOf<IRtlFieldTypeDeserializer>
...
+            const RtlFieldInfo * const *fields = base->queryFields();
+            const RtlFieldInfo *lastField = nullptr;
+            unsigned idx = 0;
+            while (fields[idx])
+            {
+                lastField = fields[idx];
+                idx++;
+            }
+            assertex(lastField);
+            switch(lastField->type->fieldType & RFTMkind)
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp]: minor: common up lastType = lastField->type

Comment (ghalliday) [reply rtl/eclrtl/rtldynfield.cpp]: Have you checked that the last field is ever non integer?
If that is true, then it would also be possible for the last field to be an integer, but not stored in the fileposition extra.

   index({unsigned id},{},'i')

File: rtl/eclrtl/rtlfield.cpp [hunk 16]
```diff
@@ -559,15 +649,15 @@ __int64 RtlSwapIntTypeInfo::getInt(const void * ptr) const
...
+        return rtlReadSwapInt(ptr, length)+bias;
 }
 
 double RtlSwapIntTypeInfo::getReal(const void * ptr) const
 {
     if (isUnsigned())
         return (double) rtlReadSwapUInt(ptr, length);
     else
-        return (double) rtlReadSwapInt(ptr, length);
+        return (double) rtlReadSwapInt(ptr, length)+bias;
```
Comment (ghalliday) [inline rtl/eclrtl/rtlfield.cpp]: should add the bias before the cast - you might lose bits otherwise...

File: thorlcr/activities/indexwrite/thindexwrite.cpp [hunk 17]
```diff
@@ -168,13 +169,21 @@ class IndexWriteActivityMaster : public CMasterActivity
...
+        // Legacy record layout info
         void * layoutMetaBuff;
         size32_t layoutMetaSize;
         if(helper->getIndexLayout(layoutMetaSize, layoutMetaBuff))
         {
             props.setPropBin("_record_layout", layoutMetaSize, layoutMetaBuff);
             rtlFree(layoutMetaBuff);
         }
+        // New record layout info
+        if (helper->queryOutputMeta() && helper->queryOutputMeta()->queryTypeInfo())
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwrite.cpp]: only if can serialize...

Comment (ghalliday) [reply thorlcr/activities/indexwrite/thindexwrite.cpp]: Should be an error if helper->queryOutputMeta() is null

File: thorlcr/activities/indexwrite/thindexwriteslave.cpp [hunk 18]
```diff
@@ -202,30 +203,23 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
         metadata->setProp("_record_ECL", helper->queryRecordECL());
 
         void * layoutMetaBuff;
```
Comment (ghalliday) [inline thorlcr/activities/indexwrite/thindexwriteslave.cpp]: unused and similar comments on serializable

File: tools/dumpkey/dumpkey.cpp [hunk 19]
```diff
@@ -136,6 +147,56 @@ int main(int argc, const char **argv)
...
+                    const RtlRecord &inrec = diskmeta->queryRecordAccessor(true);
+                    ForEachItemIn(idx, fieldNames)
+                    {
+                        unsigned fieldNum = inrec.getFieldNum(fieldNames.item(idx));
+                        if (fieldNum == (unsigned) -1)
+                            throw MakeStringException(0, "Requested output field '%s' not found", fieldNames.item(idx));
+                        fields.append(inrec.queryField(fieldNum));
+                    }
+                    fields.append(nullptr);
+                    outRecType = new RtlRecordTypeInfo(type_record, 0, fields.getArray(0)); // MORE - is length needed?
```
Comment (ghalliday) [inline tools/dumpkey/dumpkey.cpp]: length would ideally be the minimum length of the record.  It may cause problems in the future (for fixed size records) if not filled in.

File: common/thorhelper/thorxmlwrite.hpp [hunk 20]
```diff
@@ -285,43 +285,49 @@ enum XMLWriterType{WTStandard, WTEncoding, WTEncodingData64, WTJSON} ;
 thorhelper_decl CommonXmlWriter * CreateCommonXmlWriter(unsigned _flags, unsigned initialIndent=0, IXmlStreamFlusher *_flusher=NULL, XMLWriterType xmlType=WTStandard);
 thorhelper_decl IXmlWriterExt * createIXmlWriterExt(unsigned _flags, unsigned initialIndent=0, IXmlStreamFlusher *_flusher=NULL, XMLWriterType xmlType=WTStandard);
 
-class thorhelper_decl SimpleOutputWriter : implements IXmlWriter, public CInterface
+class thorhelper_decl SimpleOutputWriter : implements IXmlWriterExt, public CInterface
```
Comment (ghalliday) [inline common/thorhelper/thorxmlwrite.hpp]: was this change required for a reason?

**Review by ghalliday — COMMENTED** [review]
Various comments.  I think the indexes need a bit more thought - and a way of testing them properly.

Should we be removing getIndexLayout from eclcc if it isn't being called any more?

**Comment by ghalliday** [comment]
@richardkchapman various comments

File: common/thorhelper/thorxmlwrite.hpp (hunk 20 above)
Comment (richardkchapman) [reply common/thorhelper/thorxmlwrite.hpp]: Yes, so I could get the str() out.

File: ecl/hthor/hthor.cpp (hunk 5 above)
Comment (richardkchapman) [reply ecl/hthor/hthor.cpp]: Removed

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 10 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclhelper_dyn.cpp]: Fixed

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 9 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclhelper_dyn.cpp]: Fixed

File: rtl/eclrtl/eclhelper_dyn.cpp [hunk 21]
```diff
@@ -40,63 +40,167 @@
...
+            if (found==NotFound)
+            {
+                bool isNew;
+                unsigned insertPos = filterOffsets.bAdd(fieldOffset, compareOffsets, isNew);
+                dbgassertex(isNew);
+                filters.add(*filterSet.getClear(), insertPos);
+            }
+            else
+            {
+                filters.item(found).unionSet(filterSet);  // Debatable - would intersect be more appropriate?
```
Comment (ghalliday) [inline rtl/eclrtl/eclhelper_dyn.cpp]: The old IStringSets unionSet returns the result - rather than modifying the set => this will leak and be a nop.  (New sets modify in place.)

File: system/jhtree/jhtree.cpp [hunk 22]
```diff
@@ -660,8 +660,8 @@ class jhtree_decl CKeyLevelManager : implements IKeyManager, public CInterface
             {
                 started = true;
                 numsegs = segs.ordinality();
-                if (numsegs)
-                    segs.checkSize(keyedSize, keyName.get());
+                //if (numsegs)
```
Comment (ghalliday) [inline system/jhtree/jhtree.cpp]: Should this be uncommented or removed?

File: tools/dumpkey/dumpkey.cpp [hunk 23]
```diff
@@ -18,6 +18,10 @@
 #include "jliball.hpp"
 #include "jhtree.hpp"
 #include "ctfile.hpp"
+#include "rtlrecord.hpp"
+#include "eclhelper_dyn.hpp"
+//#include "rtldynfield.hpp"
```
Comment (ghalliday) [inline tools/dumpkey/dumpkey.cpp]: delete?

File: system/jhtree/jhtree.cpp (hunk 22 above)
Comment (richardkchapman) [reply system/jhtree/jhtree.cpp]: It should be put back once I know the keysize and can therefore add the required wildSegMonitors

File: rtl/eclrtl/eclhelper_dyn.cpp (hunk 21 above)
Comment (richardkchapman) [reply rtl/eclrtl/eclhelper_dyn.cpp]: Will fix

File: rtl/eclrtl/rtlrecord.cpp [hunk 24]
```diff
@@ -300,20 +401,28 @@ size32_t RtlRecord::getMinRecordSize() const
...
     return minSize;
 }
 
 size32_t RtlRecord::deserialize(ARowBuilder & rowBuilder, IRowDeserializerSource & in) const
 {
     size32_t offset = 0;
+    byte *conditionValues = (byte *) alloca(numIfBlocks);
     for (unsigned i = 0; i < numVarFields; i++)
     {
+        memset(conditionValues, 2, numIfBlocks);
```
Comment (ghalliday) [inline rtl/eclrtl/rtlrecord.cpp]: Should this be outside the loop?

Comment (richardkchapman) [reply rtl/eclrtl/rtlrecord.cpp]: It should be once per row. I guess it still works inside the loop but rather less efficient!

File: rtl/include/eclhelper.hpp [hunk 25]
```diff
@@ -348,9 +352,9 @@ enum RtlFieldTypeMask
     RFTMcontainsunknown     = 0x10000000,                   // contains a field of unknown type that we can't process properly
     RFTMinvalidxml          = 0x20000000,                   // cannot be called to generate xml
     RFTMhasxmlattr          = 0x40000000,                   // if specified, then xml output includes an attribute (recursive)
-    RFTMnoserialize         = 0x80000000,                   // cannot serialize this typeinfo structure (contains ifblocks, dictionaries or other nasties)
+    RFTMnoserialize         = 0x80000000,                   // cannot serialize this typeinfo structure (contains dictionaries or other nasties)
 
-    RFTMinherited           = (RFTMnoprefetch|RFTMcontainsunknown|RFTMinvalidxml|RFTMhasxmlattr|RFTMnoserialize)    // These flags are recursively set on any parent records too
+    RFTMinherited           = (RFTMcontainsifblock|RFTMnoprefetch|RFTMcontainsunknown|RFTMinvalidxml|RFTMhasxmlattr|RFTMnoserialize)    // These flags are recursively set on any parent records too
```
Comment (ghalliday) [inline rtl/include/eclhelper.hpp]: Because RFTMalien has the same value as RFTMcontainsifblock it is causing this flag to be set on any dataset with alien types.

It is also being set even if it is only a child dataset that contains the alien type - I think RFTMcontainsifblock should not be inherited from child datasets
